Ability to break on `:` for objects

yamafaktory opened the pull request:
This PR aims to solve #280.

vjeux commented:
Omg! Thanks for working on this <3

yamafaktory commented:
@vjeux I hope it actually solved it the way you were expecting it!

vjeux commented:
The examples look good to me. I'm going to run it on a bigger real-world codebase to see how it impacts things.

vjeux commented:
Here's how it looks on parts of nuclide (there's a bunch of other differences that add noise to the diff so I had to copy paste interesting pieces manually)

https://gist.github.com/vjeux/682684c96091f359359e57b0a18de078

vjeux commented:
What do you think? After looking at real code, I'm not so sure that this is an imporvement.

rhengles commented:
My 2¢: it's an improvement (only?) when the key name is very large.

yamafaktory commented:
@vjeux IMHO I see that as an improvement as it greatly improves readability but it's also a matter of taste. I'd love to get @jlongster's feedback on it 👍.

jlongster commented:
I'm not sure it's a big improvement either. I try to think about what users would expect, and I don't think I ever see the the right-hand bump down onto a new line. I think it's better to allow objects and other things that can be broken across newlines (and still look good) to be the thing that breaks.

If you have a long expression that you don't want to break down, I think it's better to tell the programmer to create a new variable:

```
const isSelected = this.state.selectedCallFrameIndex === i;
const itemClassNames = classNames({
  'nuclide-debugge-callstack-item-selected': isSelected
});
```

Additionally, you can always change the `printWidth` setting to be a little larger to give it more room. I actually think something like 90 makes sense for projects with a lot of complex code like above.

jlongster commented:
It is possible for us to easily check the length of the key name. I would be fine will allowing the value to move to the next line if the key name is past a certain length, which would help with these situations.

jlongster commented:
Note that this is common scenario though and we need to decide what we want. For example:

```
const longVariable = longExpression() ===
  1
```

Our rules allow parts of expressions to break down and indent on the next line. The reason we don't move the whole expression down is because things like this are extremely common:

```
const longVariable = {
  foo: 1
};
```

So the core problem is binary expressions where a single primitive is moved down. If it's a more complex expression, I think it looks fine:

```js
const isSelected = 
const itemClassNames = classNames({
  'nuclide-debugge-callstack-item-selected': this.state.selectedCallFrameIndex === 
    calculateIndex() + somethingElse
});
```

I like the consistency that we've found with binary expressions. Users can expect when things are broken and what they look like.

However, I agree that bumping down a simple `1` is not ideal. Maybe we could add a small edge case to binary expressions where we never bump down a primitive value? That way `this.state.selectedCallFrameIndex === 1` would never be broken up and the above code it would just all stay on one line.

yamafaktory commented:
@vjeux @jlongster so do you consider refactoring it so that it will perform a length checking in order to break or not, or just completely dropping that feature and the original issue (which is totally fine)?

jlongster commented:
@yamafaktory The issue I have with checking the length is it doesn't take into account the current column position of the expression. So it's hard to know what that length should be. If the object is indented a lot, you'd want that length small. If it's at the beginning of a line, you'd want it to be larger. I don't like encoding rules that go against the low-level printer which has a lot more information about how things are laid out. What do you think of my last comment?

yamafaktory commented:
@jlongster Oh sorry, I think I was commenting while you were also doing it at the same time... Well, regarding your last comment, this totally makes sense. Being consistent is the key, you're totally right! Adding a small exception could be the best solution 👍

lydell commented:
It's all about whether the line ends with a `(`, `[`, `{` or not. If it does, it generally looks good to break there. If not, it's nicer to move the entire expression to a new line (or wrap it in parentheses).

```js
// Does not end with ([{
const longVariable = longExpression() ===
  1;

// Move the entire expression to a new line.
const longVariable =
  longExpression() === 1;

// Also pretty nice.
const longVariable = (
  longExpression() === 1
);
```

```js
// Does not end with ([{
const itemClassNames = classNames({
  'nuclide-debugger-callstack-item-selected': this.state.selectedCallFrameIndex ===
    calculateIndex() + somethingElse
});

// Move the entire expression to a new line.
const itemClassNames = classNames({
  'nuclide-debugger-callstack-item-selected':
    this.state.selectedCallFrameIndex === calculateIndex() + somethingElse
});

// Also pretty nice.
const itemClassNames = classNames({
  'nuclide-debugger-callstack-item-selected': (
    this.state.selectedCallFrameIndex === calculateIndex() + somethingElse
  )
});

// Also notice how lines ending with `({` above look nice!
```

```js
// It's all about the ending ( here...
const height = itemHeight + Math.max(
  0,
  windowHeight - containerHeight
);
```

yamafaktory commented:
@lydell your last example is out of scope (at least regarding the initial issue #280) but wrapping the entire expression with parenthesis is also something quite common and interesting too (and I tend to like it).

jlongster commented:
@lydell Generally we can't specify rules based on the output. We have no idea what is at the "end" of a line because we have no idea if the formatter has broken up the object or not. We have no knowledge of true output in the AST print phase.

Let's try something though. Maybe this would work to see if we should move something to a new line or not:

1. We figure out the exact places that need this behavior. The two I can think of are variable declarations and object expressions (this can't be generalized because the same expression might be a function arg which shouldn't have it)
2. We check the right-hand expression and determine if it should break itself up or not.
3. If not, we add parens around it and bump to a new line.

The problem is the "determine if it should break" or not. Here's the crux of the problem with the "line ends in `[({` rule": the line itself may be so long that that object should not break, and it really should break everything onto newlines:

```js
var obj = {
  property: reallyLongExpression === anotherLongExpression + longFunctionCall({ foo: 1 })
}
```

Assume that the above line runs out of space. You would not want:

```js
var obj = {
  property: reallyLongExpression === anotherLongExpression + longFunctionCall({
    foo: 1 
  })
}
```

But you would want:

```js
var obj = {
  property: (
    reallyLongExpression === anotherLongExpression + longFunctionCall({ foo: 1 })
  )
);
```

That may not be the best example, but imaging a really long line that should break not on the function call, but somewhere else.

The point is we have no idea if a line "ends with [{(" because we don't know if the formatter broke it up or not. We need to speak in terms of AST nodes and documents. While writing this I've lost clarity and I'm not sure what we want so I need to come back to this. :)

yamafaktory commented:
@jlongster thanks a lot for the great feedback, I will try to implement your guidelines as a first draft!

jlongster commented:
@yamafaktory cool! perhaps it could be as simple as: if the value of an object property is a function expression, function call, object, or array literal, don't break to new line ever, otherwise allow it to.

yamafaktory commented:
@jlongster here we go! I added a bunch of types to be checked in order to dispatch it as breakable or not... not sure if this is the best way to do it but it's doing the job.

jlongster commented on src/printer.js:
```diff
@@ -621,7 +621,40 @@ function genericPrintNoParens(path, options, print) {
+          "StringLiteral",
+          "ThisExpression",
+          "TypeCastExpression"
+        ].reduce((stack, type) => stack || (n.value.type === type), false);
```
I think is valuable info. There might be other places we could use this logic as well. Can you move it out to a utility function at the bottom of the file? Something like `shouldPrintSameLine` or something. Not sure what the best name is.

This function just takes a node and checks these. We can also use `namedTypes.<type>.check` to check a type of node, but looking at the code I'm worried it's a bit slow, and we already don't use it many other places, so I dunno... we should at least use it for `Literal` because it works for both babylon and flow even though they define the literal types differently.  Although it looks inconsistent, the function would look like:

```js
function shouldPrintSameLine(node) {
  const type = node.type;
  return namedTypes.Literal.check(node) ||
    type === "ArrayExpression" ||
    type === "ArrayPattern" ||
    type === "ArrowFunctionExpression" ||
    ...
}
```

I'd like to avoid the `reduce` call for performance too, even if the above is a little more verbose.

jlongster commented on src/printer.js:
```diff
@@ -621,7 +621,40 @@ function genericPrintNoParens(path, options, print) {
+          "FunctionExpression",
+          "Identifier",
+          "Literal",
+          "LogicalExpression",
```
How come logical expression is here but binary expression isn't? Should probably do the same for both

jlongster replied:
(Since the original issue was that binary expressions would break weird, we probably want to allow the expression to be moved down for both binary and logical, so you should probably remove LogicalExpression here)

jlongster commented on src/printer.js:
```diff
@@ -621,7 +621,40 @@ function genericPrintNoParens(path, options, print) {
+                    : indent(options.tabWidth, concat([line, printedValue]))
+                ])
+              ),
+              noBreak ? "" : concat([line, ")"])
```
There are many different cases for `noBreak` that it'd probably be clearer to just separate it out completely. So do:

```js
if(noBreak) {
  parts.push(group(concat([ ... ]))
}
else {
  parts.push(group(concat([ ... ]))
}
```

Defining these separately might repeat one or two things, but makes it a lot more readable for what happens in that case.

jlongster commented:
Definitely looking good! This feels like the right direction to take.

yamafaktory commented:
@jlongster as suggested, I've extracted the logic inside another function (and dropped the reducer by the way). I've also removed the LogicalExpression type from the checking to be more consistent. As a consequence, I needed to update some of the core tests involving it. I finally added a new test with a long logical expression that breaks. Hope it will fit your view!

jlongster commented:
Thanks a lot for this! This looks great. So sorry for the slow review process; been refactoring a bit of code and focusing on that.

I made a few tweaks as seen here: https://github.com/jlongster/prettier/pull/314/commits/9bf2541e0cac266d51d55060adbff746d4c17e04. We've removed `multilineGroup` in a recent commit, and you also only need to print parens if you are breaking. Otherwise this is awesome! Thanks a lot!

jlongster commented:
There seems to be a bug on github. It's not running the tests on the right branch. They pass fine locally for me. I'm going to merge this and watch master closely to see if it's failing.

yamafaktory commented:
@jlongster Absolutely no problem! You're doing an awesome job (and so do the other committers of course) and this is a pleasure to try to help ( I'm definitely waiting for more maturity to integrate prettier in my company dev workflow as a _de facto_ tool in the pipeline 👍).

vjeux commented:
FYI, this introduces a regression for trailing commas: https://github.com/jlongster/prettier/issues/461

vjeux commented on tests/logical/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -901,14 +901,14 @@ function logical12b(y: number): number {
-    { x: \"foo\" && \"bar\" },
-    { x: \"foo\" && \"bar\" },
-    { x: \"foo\" && \"bar\" }
+    { x: x && \"bar\"  },
```
There is one too many space here

yamafaktory replied:
Oh, good catch 👍
